libvmi/os/linux/core.c

```diff
@@ -28,10 +28,75 @@
 #include "config/config_parser.h"
 #include "driver/driver_wrapper.h"
 #include "os/linux/linux.h"
+#include <glib.h>
+
 
 void linux_read_config_ghashtable_entries(char* key, gpointer value,
         vmi_instance_t vmi);
 
+static status_t init_task_kaslr_test(vmi_instance_t vmi, addr_t page_vaddr);
+
+static status_t init_kaslr(vmi_instance_t vmi);
+
+static status_t brute_force_find_kern_mem (vmi_instance_t vmi);
+
+static status_t verify_linux_paging (vmi_instance_t vmi);
+
+#if defined(X86_64)
+/* Is the given physical address an Intel x64 page directory structure? */
+static bool is_x86_64_pd (vmi_instance_t vmi, addr_t pa)
+{
+    bool rc = false;
+    status_t status = VMI_FAILURE;
+
+#define PD_ENTRIES (VMI_PS_4KB / sizeof(uint64_t))
+    uint64_t page[PD_ENTRIES];
+    addr_t maxframe = vmi_get_max_physical_address (vmi) >> 12;
+
+    status = vmi_read_pa (vmi, pa, sizeof(page), (void *)page, NULL);
+    if (VMI_FAILURE == status)
+        return false;
+
+    for (size_t i = 0; i < PD_ENTRIES; ++i) {
+        if (0 == (page[i] & 1)) /* present bit is cleared, so skip */
+            continue;
+
+        /* present bit is 1, and ... */
+        addr_t gfn = page[i] >> 12;
+        if (0 == gfn || gfn > maxframe) {
+            /* ... this is not a valid GFN. Fail the page. */
+            rc = false;
+            goto exit;
+        }
+
+        /* ... the GFN looks reasonable */
+        rc = true;
+    }
+
+exit:
+    return rc;
+}
+#endif
+
+
+/*
+ * Identifies page directories in memory. This saves multiple minutes
+ * off a brute force search for the KPDB and KASLR offset.
+ */
+static GSList * find_page_directories (vmi_instance_t vmi)
+{
+    GSList * list = NULL;
+
+    for (addr_t candidate = 0x1000; candidate < vmi_get_max_physical_address (vmi); candidate += VMI_PS_4KB) {
+#if defined(X86_64)
+        if (is_x86_64_pd (vmi, candidate)) {
+            list = g_slist_prepend (list, (gpointer) candidate);
```

Comment:
You should probably be using [`GUINT_TO_POINTER`](https://developer.gnome.org/glib/stable/glib-Type-Conversion-Macros.html#GUINT-TO-POINTER:CAPS) here

Reply:
Even then this might be problematic when LibVMI is compiled as 32-bit and you try to stuff an `addr_t` in there which is 64-bit.

Reply:
It should be OK. As is, the list is only populated on 64-bit.

Reply:
Yea, but the compiler is complaining about it regardless. That's why Jenkins is failing the PR.

Reply:
Also, just keep in mind that there is no requirement on LibVMI's side that it needs to be compiled as 64-bit in order to introspect a 64-bit VM. It could be compiled as 32-bit and introspect a 64-bit VM, which is where the problem I describe comes up.

Reply:
I agree - given the nature of our codebase, it's always preferable to use fully explicit types.

Reply:
Ah, good point. So there's an additional problem here, because the check for whether to use this heuristic should be based on the target architecture (e.g. `vmi_get_address_width()`). But how do I tell whether the target is Intel or ARM?

Reply:
See https://github.com/libvmi/libvmi/blob/master/libvmi/libvmi.h#L848